File: site/src/types/browse.ts
export interface RawBrowseResponse {
  files: RawBrowseFile[];
  project_resolution?: ProjectResolution;
}
export interface ProjectResolution {
  project_id?: string;
  version_id?: number;
}

export interface RawFile {
  name: string;
  mime_type: string;
  file_size: number;
  modified: string;
  created: string;
}
export interface RawDirectory {
  name: string;
  number_of_files: number;
}

export type RawBrowseFile =
  | {
      type: "File";
      value: RawFile;
    }
  | {
      type: "Directory";
      value: RawDirectory;
    };
export function fixCurrentPath(path: string): string {
  let newPath = path;
  if (newPath.startsWith("/")) {
    newPath = newPath.substring(1);
  }
  if (newPath.endsWith("/")) {
    newPath = newPath.substring(0, newPath.length - 1);
  }
  return newPath;
}

File: crates/core/src/repository/browse.rs
use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::storage::{FileHashes, SerdeMime};

use super::project::ProjectResolution;

#[derive(Debug, Serialize, Deserialize, Clone, ToSchema)]
#[serde(tag = "type", content = "value")]
pub enum BrowseFile {
    File {
        name: String,
        file_size: u64,
        mime_type: Option<SerdeMime>,
        file_hash: FileHashes,
        /// Last time it was modified.
        modified: DateTime<FixedOffset>,
        /// The first time it was created.
        created: DateTime<FixedOffset>,
    },
    Directory {
        name: String,
        number_of_files: usize,
        //modified: DateTime<FixedOffset>,
    },
}
#[derive(Debug, Serialize, Deserialize, Clone, ToSchema)]
pub struct BrowseResponse {
    pub files: Vec<BrowseFile>,
    /// The project contained in the path
    ///
    /// None if the check_for_project is false
    pub project_resolution: Option<ProjectResolution>,
}
